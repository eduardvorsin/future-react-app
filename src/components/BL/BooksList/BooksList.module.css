.books-list {
	display: grid;
	grid-auto-rows: min-content;
	grid-template-columns: 1fr;
	gap: 20px 20px;
	overflow: hidden;
	max-width: 350px;
	margin: 0 auto;
}

.books-list__item {
	/* .books-list__item styles */
}

@media(min-width:576px) {
	.books-list {
		gap: 40px 20px;
		max-width: 100%;
		grid-template-columns: 1fr 1fr;
	}
}

@media(min-width:768px) {
	.books-list {
		grid-template-columns: repeat(3, 1fr);
	}
}

@media(min-width:992px) {
	.books-list {
		grid-template-columns: repeat(4, 1fr);
	}
}