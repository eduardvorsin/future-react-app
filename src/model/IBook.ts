import { AccessInfo, SaleInfo, VolumeInfo } from '../API/bookTypes';

export type IBook = VolumeInfo & AccessInfo & SaleInfo;

export type IBookPartial =
  { id: string } &
  Pick<VolumeInfo, 'authors' | 'maturityRating' | 'title' | 'imageLinks'> & Pick<SaleInfo, 'retailPrice'>;
