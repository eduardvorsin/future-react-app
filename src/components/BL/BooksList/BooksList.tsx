import React from 'react';
import { useTranslation } from 'react-i18next';
import { IBookPartial } from '../../../model/IBook';
import Card from '../../UI/Card/Card';
import classes from './BooksList.module.css';

type BooksListProps = {
  className?: string,
  data: IBookPartial[],
}

const BooksList: React.FC<BooksListProps> = ({
  className,
  data,
}) => {
  const { t } = useTranslation();
  const bookListClasses = [
    className,
    classes['books-list'],
  ].join(' ');

  return (
    <div
      className={bookListClasses}
    >
      {data.map((item) => (
        <Card
          id={item.id}
          key={item.id}
          maturityRating={item.maturityRating}
          className={classes['books-list__item']}
          price={item?.retailPrice}
          title={item.title}
          authors={item?.authors}
          src={item?.imageLinks?.smallThumbnail}
          alt={`${item.title} ${t('bookCover')}`}
        />
      ))}
    </div>
  );
};

export default BooksList;
