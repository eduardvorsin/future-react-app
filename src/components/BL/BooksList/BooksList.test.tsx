import React from 'react';
import { render, screen } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';
import BooksList from './BooksList';
import { IBookPartial } from '../../../model/IBook';

const mockData: IBookPartial[] = [{
  id: '123',
  clientId: '123',
  authors: ['authors'],
  maturityRating: 'NOT_MATURE',
  title: 'title',
  imageLinks: {
    smallThumbnail: '',
    thumbnail: '',
  },
  retailPrice: {
    amount: 12,
    currencyCode: 'USD',
  },
}];

const router = createMemoryRouter([{
  path: '/',
  element: (
    <BooksList
      data={mockData}
      testId='test-book-list'
    />
  ),
}], { initialEntries: ['/'] });

const routerWithEmptyData = createMemoryRouter([{
  path: '/',
  element: (
    <BooksList
      data={[]}
      testId='test-book-list'
    />
  ),
}], { initialEntries: ['/'] });

describe('BooksList tests', () => {
  it('is rendered correctly', () => {
    render(
      <RouterProvider router={router} />,
    );

    expect(screen.getByTestId<HTMLDivElement>('test-book-list')).toBeInTheDocument();
  });

  it('if an empty array is passed, then the cards should not be rendered', () => {
    render(
      <RouterProvider router={routerWithEmptyData} />,
    );

    expect(screen.queryByText<HTMLParagraphElement>(/authors/i)).not.toBeInTheDocument();
    expect(screen.queryByText<HTMLParagraphElement>(/learn more/i)).not.toBeInTheDocument();
    expect(screen.queryByText<HTMLSpanElement>(/price/i)).not.toBeInTheDocument();
  });

  it('is a basic snapshot', () => {
    render(
      <RouterProvider router={router} />,
    );

    expect(screen.getByTestId<HTMLDivElement>('test-book-list')).toMatchSnapshot();
  });

  it('is a snapshot with an empty data array', () => {
    render(
      <RouterProvider router={routerWithEmptyData} />,
    );

    expect(screen.getByTestId<HTMLDivElement>('test-book-list')).toMatchSnapshot();
  });
});
