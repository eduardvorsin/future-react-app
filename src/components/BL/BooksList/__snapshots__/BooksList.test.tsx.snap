// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksList tests is a basic snapshot 1`] = `
<div
  class=" books-list"
  data-testid="test-book-list"
>
  <article
    class="books-list__item card"
  >
    <div
      class="card__img-wrapper "
      data-testid="image-wrapper"
    >
      <img
        alt="title book cover"
        class="card__img"
        height="185"
        src=""
        width="128"
      />
      <a
        href="/books/123"
      >
        Learn more
      </a>
    </div>
    <div
      class="card__content"
    >
      <p
        class="card__price"
      >
        <span
          class="sr-only"
        >
          Book price
          :
        </span>
        $12
      </p>
      <h3
        class="title card__title title--dark title--level-6"
      >
        <a
          href="/books/123"
        >
          title
        </a>
      </h3>
      <p
        class="card__authors"
      >
        Authors
        : 
        authors
      </p>
    </div>
    <a
      aria-disabled="false"
      class="link card__more-link  link--secondary"
      href="/books/123"
    >
      Learn more
    </a>
  </article>
</div>
`;

exports[`BooksList tests is a snapshot with an empty data array 1`] = `
<div
  class=" books-list"
  data-testid="test-book-list"
/>
`;
